File: packages/core/ts/src/events.ts
// botiva agent events — the common language between runtimes (LangGraph,
// LangChain, custom code, any language) and the ConversationEngine.
//
// A turn produces this event sequence:
//   run_started → (tool_call | message | genui | interrupt)* → run_finished | run_error
//
// tool_call: emitted multiple times with the same id (running → completed/error);
// clients upsert by id (chativa renders this as the ToolCallActivity strip).
// interrupt: human-in-the-loop — the run pauses, the user's next message is fed
// back to the runtime as `resume`.
// genui: mounts/updates a component registered in the client GenUIRegistry.

export type ToolCallStatus = "running" | "completed" | "error";

/** Mirrors the chativa ToolCall entity. */
export interface ToolCall {
    id: string;
    name: string;
    status: ToolCallStatus;
    params?: unknown;
    result?: unknown;
    error?: string;
    startedAt?: number;
    endedAt?: number;
}

/** Mirrors the chativa MessageAction chip. */
export interface MessageAction {
    label: string;
    value?: string;
}

/** Mirrors the chativa AIChunk. */
export type GenUIChunk =
    | { type: "ui"; component: string; props?: Record<string, unknown>; id?: string | number }
    | { type: "text"; content: string; id?: string | number }
    | { type: "event"; name: string; payload?: unknown; id?: string | number };

/** Recommended interrupt payload shape — rendered as a question + action chips. */
export interface InterruptQuestion {
    question?: string;
    message?: string;
    options?: Array<string | MessageAction>;
    [key: string]: unknown;
}

export type AgentEvent =
    | { type: "run_started" }
    | { type: "run_finished" }
    | { type: "run_error"; error: string }
    | { type: "message"; text: string; actions?: MessageAction[] }
    | { type: "tool_call"; toolCall: ToolCall }
    | { type: "interrupt"; payload: unknown; id?: string | null }
    | { type: "genui"; chunk: GenUIChunk; streamId?: string; done?: boolean }
    | { type: "busy" };

// ── factories ────────────────────────────────────────────────────────────────

export const runStarted = (): AgentEvent => ({ type: "run_started" });
export const runFinished = (): AgentEvent => ({ type: "run_finished" });
export const runError = (error: unknown): AgentEvent => ({ type: "run_error", error: String(error) });

export const message = (text: string, actions?: MessageAction[]): AgentEvent =>
    actions ? { type: "message", text, actions } : { type: "message", text };

export const busy = (): AgentEvent => ({ type: "busy" });

export const toolCall = (
    id: string,
    name: string,
    status: ToolCallStatus,
    extra: Partial<Omit<ToolCall, "id" | "name" | "status">> = {},
): AgentEvent => ({ type: "tool_call", toolCall: { id, name, status, ...extra } });

/** Recommended payload shape: { question, options? } → rendered as action chips. */
export const interrupt = (payload: unknown, id: string | null = null): AgentEvent => ({
    type: "interrupt",
    payload,
    id,
});

/**
 * Generative UI chunk. `streamId` is optional — the engine groups all chunks
 * of a turn under one stream and closes it automatically.
 */
export const genui = (chunk: GenUIChunk, opts: { streamId?: string; done?: boolean } = {}): AgentEvent => ({
    type: "genui",
    chunk,
    ...(opts.streamId !== undefined ? { streamId: opts.streamId } : {}),
    ...(opts.done !== undefined ? { done: opts.done } : {}),
});

/** Shorthand: mount a client-side component (chativa GenUIRegistry) in the chat. */
export const ui = (component: string, props: Record<string, unknown> = {}): AgentEvent =>
    genui({ type: "ui", component, props });

File: packages/core/go/events.go
// Package botiva is the Go reference port of the botiva conversation
// framework (Botiva Wire Protocol v1). Signatures mirror @botiva/core —
// see PROTOCOL.md §8 for the cross-language contract.
package botiva

// ToolCall mirrors the chativa ToolCall entity.
type ToolCall struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Status    string `json:"status"` // running | completed | error
	Params    any    `json:"params,omitempty"`
	Result    any    `json:"result,omitempty"`
	Error     string `json:"error,omitempty"`
	StartedAt int64  `json:"startedAt,omitempty"`
	EndedAt   int64  `json:"endedAt,omitempty"`
}

// MessageAction mirrors the chativa MessageAction chip.
type MessageAction struct {
	Label string `json:"label"`
	Value string `json:"value,omitempty"`
}

// GenUIChunk mirrors the chativa AIChunk ("ui" | "text" | "event").
type GenUIChunk struct {
	Type      string         `json:"type"`
	Component string         `json:"component,omitempty"`
	Props     map[string]any `json:"props,omitempty"`
	Content   string         `json:"content,omitempty"`
	Name      string         `json:"name,omitempty"`
	Payload   any            `json:"payload,omitempty"`
	ID        any            `json:"id,omitempty"`
}

// AgentEvent is the runtime → engine event. One struct with a Type
// discriminator keeps the wire mapping identical to the TS union.
type AgentEvent struct {
	Type     string          `json:"type"` // run_started|run_finished|run_error|message|tool_call|interrupt|genui|busy
	Text     string          `json:"text,omitempty"`
	Actions  []MessageAction `json:"actions,omitempty"`
	ToolCall *ToolCall       `json:"toolCall,omitempty"`
	Payload  any             `json:"payload,omitempty"`
	ID       string          `json:"id,omitempty"`
	Chunk    *GenUIChunk     `json:"chunk,omitempty"`
	StreamID string          `json:"streamId,omitempty"`
	Done     bool            `json:"done,omitempty"`
	Error    string          `json:"error,omitempty"`
}

// ── factories (mirror @botiva/core) ─────────────────────────────────────────

func RunStarted() AgentEvent          { return AgentEvent{Type: "run_started"} }
func RunFinished() AgentEvent         { return AgentEvent{Type: "run_finished"} }
func RunError(err string) AgentEvent  { return AgentEvent{Type: "run_error", Error: err} }
func Busy() AgentEvent                { return AgentEvent{Type: "busy"} }
func Message(text string) AgentEvent  { return AgentEvent{Type: "message", Text: text} }

func ToolCallEvent(tc ToolCall) AgentEvent { return AgentEvent{Type: "tool_call", ToolCall: &tc} }

// Interrupt: recommended payload {question, options} → rendered as chips.
func Interrupt(payload any, id string) AgentEvent {
	return AgentEvent{Type: "interrupt", Payload: payload, ID: id}
}

func GenUI(chunk GenUIChunk) AgentEvent { return AgentEvent{Type: "genui", Chunk: &chunk} }

// UI mounts a client-registered component (chativa GenUIRegistry).
func UI(component string, props map[string]any) AgentEvent {
	return GenUI(GenUIChunk{Type: "ui", Component: component, Props: props})
}
